pkg/etcd-backup-operator/controller/s3backup.go

```diff
@@ -0,0 +1,229 @@
+	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+	defer rc.Close()
+	if err != nil {
+		return fmt.Errorf("failed to receive snapshot: (%v)", err)
+	}
+	err = s.saveToS3(resp.Version, rev, rc)
+	if err != nil {
+		return fmt.Errorf("failed to save snapshot: (%v)", err)
+	}
+	return nil
+}
+
+func (s *s3Backup) saveToS3(version string, rev int64, rc io.ReadCloser) error {
+	secret, err := s.kubecli.Core().Secrets(s.namespace).Get(s.awsSecret, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to retrieve secret %v: (%v)", s.awsSecret, err)
+	}
+	s3u, err := newS3Uploader(s.s3bucket, s.prefix, secret)
+	if err != nil {
+		return fmt.Errorf("failed to create s3 uploader: (%v)", err)
+	}
+	key := path.Join(s.prefix, makeBackupName(version, rev))
+	ui := &s3manager.UploadInput{
+		Bucket: aws.String(s.s3bucket),
+		Key:    aws.String(key),
+		Body:   rc,
+	}
+	_, err = s3u.Upload(ui)
+	if err != nil {
+		return fmt.Errorf("failed to upload snapshot to s3: (%v)", err)
+	}
+	return nil
+}
+
+func makeBackupName(ver string, rev int64) string {
+	return fmt.Sprintf("%s_%016x_%s", ver, rev, backupFilenameSuffix)
+}
+
+// NewS3Uploader returns a Uploader that is configured with proper aws config and credentials derived from Kubernetes secret.
+func newS3Uploader(bucket, prefix string, secret *v1.Secret) (*s3manager.Uploader, error) {
+	so, err := newAWSConfig(secret, prefix)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup aws config: (%v)", err)
+	}
+	return newFromSessionOpt(bucket, prefix, *so)
+}
+
+func newAWSConfig(secret *v1.Secret, prefix string) (*session.Options, error) {
+	dir := filepath.Join(operatorRoot, "aws", prefix)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return nil, fmt.Errorf("failed to create dir %v: (%v)", dir, err)
+	}
+
+	options := &session.Options{}
+	options.SharedConfigState = session.SharedConfigEnable
+
+	creds := secret.Data[api.AWSSecretCredentialsFileName]
+	if len(creds) != 0 {
+		credsFile := path.Join(dir, "credentials")
+		err := ioutil.WriteFile(credsFile, creds, 0600)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write credentials file: (%v)", err)
+		}
+		options.SharedConfigFiles = append(options.SharedConfigFiles, credsFile)
+	}
+
+	config := secret.Data[api.AWSSecretConfigFileName]
+	if config != nil {
+		configFile := path.Join(dir, "config")
+		err := ioutil.WriteFile(configFile, config, 0600)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write config file: (%v)", err)
+		}
+		options.SharedConfigFiles = append(options.SharedConfigFiles, configFile)
+	}
+
+	return options, nil
+}
+
+func newFromSessionOpt(bucket, prefix string, so session.Options) (*s3manager.Uploader, error) {
+	sess, err := session.NewSessionWithOptions(so)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new AWS session: (%v)", err)
+	}
+	return s3manager.NewUploader(sess), nil
+}
+
+func getMemberWithMaxRev(pods []*v1.Pod, tc *tls.Config) (*etcdutil.Member, int64) {
```

Comment:
This is duplicate code.
Please refactor and reuse.
Same for every other potential duplicate code.

Reply:
@hongchaodeng i think this one is fine since the other `getMemberWithMaxRev` will be deleted.

Reply:
I see what you are trying to do...

Please refactor and reuse the codebase in pkg/backup/ in both places.

Reply:
kk